Adding CustomFn layer

luketpeterson opened the pull request:
Adding CustomFn layer, where any Fn type, (function or immutable closure) can provide config values.

This is useful to support additional data sources, for example to get values from a database or from an unsupported file format, e.g. hjson (https://crates.io/crates/deser-hjson)

Thank you.

luketpeterson commented:
Here is a code excerpt from the test, to see the usage:

```rust
    #[config]
    #[derive(Debug, Default)]
    struct TestCfg {
        test: String,
        another: usize,
    }

    std::env::set_var("ANOTHER", "5");

    let func = || {
        to_value(TestCfg{test: "from func".to_string(), another: 25}).unwrap()
    };

    let prio = vec![Layer::CustomFn(func.into()), Layer::Env(None)];
    let config = TestCfg::with_layers(&prio).unwrap();
    assert_eq!(config.test, String::from("from func"));
    assert_eq!(config.another, 5usize);
```

bnjjj commented:
Hey @luketpeterson it looks great ! Thank you ! Could you just document it in the README [here](https://github.com/bnjjj/twelf/blob/master/README.md?plain=1#L17) to let users know that layer exists. Thanks a lot

bnjjj commented:
(tests are failing too)

luketpeterson commented:
The test issue is that the set of features enabled by the github action (equivalent to running `cargo test --features "clap, clap_rs, default, dhall, env, envy, ini, json, serde_dhall, serde_ini, serde_yaml, toml, toml_rs, yaml"`) doesn't include the new feature I added, "custom_fn", however it's still running the custom_fn test.

However, the same situation exists for nearly all the optional features.  For example, invoking `cargo test` without the `toml` feature specified causes the `toml` tests to fail.

luketpeterson commented:
I could conditionally exclude the test (as is done in `tests/default.rs`) however, this seems to demote the test to the status of sample code (ie never actually run with `cargo test`) on account of this: https://users.rust-lang.org/t/feature-based-conditional-excludes-test-from-cargo-test-inventory/95924

luketpeterson commented:
Seems the way I (and the tests/default.rs) file were doing conditional compilation was just broken.  I fixed that, and added feature gates to all tests, so now `cargo test` should pass regardless of which combination of features are enabled.

luketpeterson commented:
Committed README updates, also synchronizing README content across sub-crates within the repo.

bnjjj commented:
Thanks a lot. Released on `0.12.0`
